### PR TITLE
initial commit, kind of works, needs more metrics

### PR DESCRIPTION
**Pull request by godber**
This works ok, needs more metrics added.

### CODE REVIEW

**Comment by godber** [comment]
Here are the set of metrics currently exposed

```
# HELP teraslice_controller_slicers_count Number of execution controllers (slicers) running for this execution.
# HELP teraslice_controller_slices_failed Number of slices failed.
# HELP teraslice_controller_slices_processed Number of slices processed.
# HELP teraslice_controller_slices_queued Number of slices queued for processing.
# HELP teraslice_controller_workers_active Number of Teraslice workers actively processing slices.
# HELP teraslice_controller_workers_available Number of Teraslice workers running and waiting for work.
# HELP teraslice_controller_workers_disconnected Total number of Teraslice workers that have disconnected from execution controller for this job.
# HELP teraslice_controller_workers_joined Total number of Teraslice workers that have joined the execution controller for this job.
# HELP teraslice_controller_workers_reconnected Total number of Teraslice workers that have reconnected to the execution controller for this job.
# HELP teraslice_execution_cpu_limit CPU core limit for a Teraslice worker container.
# HELP teraslice_execution_cpu_request Requested number of CPU cores for a Teraslice worker container.
# HELP teraslice_execution_created_timestamp_seconds Execution creation time.
# HELP teraslice_execution_memory_limit Memory limit for Teraslice a worker container.
# HELP teraslice_execution_memory_request Requested amount of memory for a Teraslice worker container.
# HELP teraslice_execution_slicers Number of slicers defined on the execution.
# HELP teraslice_execution_status Current status of the Teraslice execution.
# HELP teraslice_execution_updated_timestamp_seconds Execution update time.
# HELP teraslice_execution_workers Number of workers defined on the execution.  Note that the number of actual workers can differ from this value.
# HELP teraslice_master_info Information about the teraslice master node.
# HELP teraslice_query_duration Total time to complete the named query, in ms.
```

**Comment by godber** [comment]
I am going to see how much the query on cluster state costs relative to the other queries.  If it's not too bad, I think I'd like to add a metric that shows the docker image/version for each execution.

**Comment by godber** [comment]
Here's the new set of metrics.

```
# HELP teraslice_controller_workers_active Number of Teraslice workers actively processing slices.
# HELP teraslice_controller_workers_available Number of Teraslice workers running and waiting for work.
# HELP teraslice_controller_workers_joined Total number of Teraslice workers that have joined the execution controller for this job.
# HELP teraslice_controller_workers_reconnected Total number of Teraslice workers that have reconnected to the execution controller for this job.
# HELP teraslice_controller_workers_disconnected Total number of Teraslice workers that have disconnected from execution controller for this job.
# HELP teraslice_master_info Information about the Teraslice master node.
# HELP teraslice_execution_info Information about Teraslice execution.
# HELP teraslice_controller_slicers_count Number of execution controllers (slicers) running for this execution.
# HELP teraslice_query_duration Total time to complete the named query, in ms.
# HELP teraslice_execution_cpu_limit CPU core limit for a Teraslice worker container.
# HELP teraslice_execution_cpu_request Requested number of CPU cores for a Teraslice worker container.
# HELP teraslice_execution_memory_limit Memory limit for Teraslice a worker container.
# HELP teraslice_execution_memory_request Requested amount of memory for a Teraslice worker container.
# HELP teraslice_execution_status Current status of the Teraslice execution.
# HELP teraslice_controller_slices_processed Number of slices processed.
# HELP teraslice_controller_slices_failed Number of slices failed.
# HELP teraslice_controller_slices_queued Number of slices queued for processing.
# HELP teraslice_execution_created_timestamp_seconds Execution creation time.
# HELP teraslice_execution_updated_timestamp_seconds Execution update time.
# HELP teraslice_execution_slicers Number of slicers defined on the execution.
# HELP teraslice_execution_workers Number of workers defined on the execution.  Note that the number of actual workers can differ from this value.
```

**Comment by godber** [comment]
This works fairly well I think.  There are probably some sins in here:

* too many `any`s
* do some restructuring
  * get the interfaces out in their own file
  * maybe get the metrics out too
* lots of blocking for loops

But I think it should work.  Maybe @peterdemartini or @jsnoble could prioritize these baddies and I could do some of them, but I'd like to wrap it up soon.

Oh, I need to do packaging too ... at least a dockerfile.

File: src/index.ts [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+          jobId: worker.job_id,
+          image: worker.image,
+        };
+      }
+    }
+  }
+
+  // eslint-disable-next-line no-restricted-syntax
+  for (const [, execution] of Object.entries(executions)) {
+    const regex = /.*:(.*)_.*/g;
```
Comment (peterdemartini) [inline src/index.ts]: What is this doing?

Comment (godber) [reply src/index.ts]: Extracting what is assumed to be a version from an image like this:

```
terascope/teraslice:v0.71.1_XXXX
```

Now that you call it out ... that is an internal tag structure isn't it.  Ultimately, if it doesn't parse, I just want it to be an empty string.  I could probably massage that regex a bit so it's more flexible.

Oh, and to get `matchAll` to work, I had to set `"target": "esnext", ` in the `tsconfig.json`.

**Comment by godber** [comment]
Oh, and of course I have zero tests here.  Which I regretted about half way through ... which sounds right :)

File: src/teraslice-stats.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            this.state = state.data;
+            await this.updateExecutions();
+
+            this.queryDuration.info = info.queryDuration;
+            this.queryDuration.jobs = jobs.queryDuration;
+            this.queryDuration.controllers = controllers.queryDuration;
+            this.queryDuration.state = state.queryDuration;
+        };
+        await run().catch((err) => {
+            throw new Error(`Error caught on run() ${this.baseUrl}: ${err}`);
```
Comment (peterdemartini) [inline src/teraslice-stats.ts]: Typically this kind of error catching doesn't help you find the problem, I think removing the catch here, the actual error stack will make the stack much easier to figure out what happened.

**Review by peterdemartini — COMMENTED** [review]
This is looking much better, just one additional comment

**Comment by godber** [comment]
After switching to a multi image build, I've got the docker image below 300MB.

```
teraslice-exporter                          v0.1.0                      0dc145eebc83        About an hour ago   296MB
```

I consider this done.

File: src/teraslice-stats.ts (hunk 2 above)
Comment (godber) [reply src/teraslice-stats.ts]: I removed the catch.

**Pull request merged**